File: lego-osek-master/crv/systems/lego-osek/properties/steering_sign/SteeringSignMon.hs
module Main where

import Language.Copilot
import Copilot.Compile.C99
import Data.Int (Int32)

import Prelude hiding ((&&), (||), not, (==), (/=), (<=), (>=), (>), (<))

fr_steering_angle :: Stream Int32
fr_steering_angle = extern "fr_steering_angle" Nothing

fr_cmd_steer :: Stream Int32
fr_cmd_steer = extern "fr_cmd_steer" Nothing

fr_motor_steer :: Stream Int32
fr_motor_steer = extern "fr_motor_steer" Nothing

spec :: Spec
spec = do
  let steeringErr = ((fr_cmd_steer * 40) `Language.Copilot.div` 100) - fr_steering_angle
  let posOk = (steeringErr <= 0) || (fr_motor_steer > 0)
  let negOk = (steeringErr >= 0) || (fr_motor_steer < 0)
  let zeroOk = (steeringErr /= 0) || (fr_motor_steer == 0)
  let p = posOk && negOk && zeroOk
  trigger "steering_sign_violation" (not p) []

main :: IO ()
main = reify spec >>= compile "steering_sign_mon"

File: lego-osek-master/crv/systems/lego-osek/properties/steering_sign/steering_sign_mon.c
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

#include "steering_sign_mon_types.h"
#include "steering_sign_mon.h"

static int32_t fr_cmd_steer_cpy;
static int32_t fr_steering_angle_cpy;
static int32_t fr_motor_steer_cpy;

static bool steering_sign_violation_0_guard(void) {
  return !(((((((fr_cmd_steer_cpy) * ((int32_t)(40))) / ((int32_t)(100))) - (fr_steering_angle_cpy)) <= ((int32_t)(0))) || ((fr_motor_steer_cpy) > ((int32_t)(0)))) && (((((((fr_cmd_steer_cpy) * ((int32_t)(40))) / ((int32_t)(100))) - (fr_steering_angle_cpy)) >= ((int32_t)(0))) || ((fr_motor_steer_cpy) < ((int32_t)(0)))) && ((((((fr_cmd_steer_cpy) * ((int32_t)(40))) / ((int32_t)(100))) - (fr_steering_angle_cpy)) != ((int32_t)(0))) || ((fr_motor_steer_cpy) == ((int32_t)(0))))));
}

void step(void) {
  (fr_cmd_steer_cpy) = (fr_cmd_steer);
  (fr_steering_angle_cpy) = (fr_steering_angle);
  (fr_motor_steer_cpy) = (fr_motor_steer);
  if ((steering_sign_violation_0_guard)()) {
    {(steering_sign_violation)();}
  };
}
